app/src/org/commcare/android/analytics/TimedStatsTracker.java

```diff
@@ -0,0 +1,112 @@
+        editor.putString(KEY_LAST_FORM_NAME, formTitle).
+                putLong(KEY_LAST_FORM_START_TIME, currentTime());
+        editor.commit();
+    }
+
+    public static void registerExitForm(String formTitle) {
+        SharedPreferences prefs = CommCareApplication._().getCurrentApp().getAppPreferences();
+        long enterTime = prefs.getLong(KEY_LAST_FORM_START_TIME, -1);
+        if (enterTime != -1) {
+            String formLastEntered = prefs.getString(KEY_LAST_FORM_NAME, "");
+            if (formLastEntered.equals(formTitle)) {
+                reportTimedEvent(GoogleAnalyticsFields.ACTION_TIME_IN_A_FORM, formTitle,
+                        computeElapsedTimeInSeconds(enterTime, currentTime()));
+            } else {
+                Log.i(TAG, "Attempting to report exit form time for a different form than the " +
+                        "last form logged as entered");
+            }
+        } else {
+            Log.i(TAG, "Attempting to report exit form time when there was no start form time " +
+                    "saved in prefs");
+        }
+    }
+
+    public static void registerStartSession() {
+        SharedPreferences.Editor editor =
+                CommCareApplication._().getCurrentApp().getAppPreferences().edit();
+        String currentUserId = CommCareApplication._().getCurrentUserId();
+        if (!"".equals(currentUserId)) {
+            editor.putLong(KEY_LAST_SESSION_START_TIME, currentTime())
+                    .putString(KEY_LAST_LOGGED_IN_USER, currentUserId);
+            editor.commit();
+        } else {
+            Log.i(TAG, "Attempting to report starting a session with no logged in user available");
+        }
+    }
+
+    public static void registerEndSession(String loggedOutUser) {
+        if (loggedOutUser == null || "".equals(loggedOutUser)) {
+            Log.i(TAG, "Attempting to report ending a session with no logged in user available");
+            return;
+        }
+        SharedPreferences prefs = CommCareApplication._().getCurrentApp().getAppPreferences();
+        String lastLoggedInUser = prefs.getString(KEY_LAST_LOGGED_IN_USER, "");
+        if (!"".equals(lastLoggedInUser)) {
+            if (lastLoggedInUser.equals(loggedOutUser)) {
+                long startTime = prefs.getLong(KEY_LAST_SESSION_START_TIME, -1);
+                reportTimedEvent(GoogleAnalyticsFields.ACTION_SESSION_LENGTH,
+                        computeElapsedTimeInSeconds(startTime, currentTime()));
+            } else {
+                Log.i(TAG, "Attempting to report ending a session for a different user than the " +
+                        "last user reported as logged in");
+            }
+        } else {
+            Log.i(TAG, "Attempting to report ending a session when there is no last logged in " +
+                    "user in prefs");
+        }
+    }
+
+    private static int computeElapsedTimeInSeconds(long startTime, long endTime) {
+        return (int) (endTime - startTime / 1000);
```

Comment:
adding an extra pair of around `endTime - startTime` would be nice for those of us who never memorized arithmetic operator precedence, like me :)